syft/core/frameworks/torch/hook.py

```diff
@@ -0,0 +1,298 @@
+
+    #         Args:
+    #         workers: string (or sequence) containing IPFS address(es)
+    #             of worker node(s).
+    #         """
+
+    #         # makes singleton, if needed
+    #         workers = hook_self.local_worker._check_workers(self, workers)
+
+    #         for worker in workers:
+    #             hook_self.local_worker.send_obj(self,
+    #                                             worker)
+
+
+    #     setattr(tensorvar_type, 'send_', send_)
+    #     setattr(tensorvar_type, 'send', send_)
+                
+    def _hook_LocalTensor(self, tensor_type):
+        
+        # iterate through all methods and tell them to call the native function
+        # on self.child
+        for attr in self.to_auto_overload[tensor_type]:
+
+            lit = getattr(tensor_type, attr)
+            passer = utils.pass_method_args(lit)
+            new_attr = self._forward_call_to_children(passer, attr, call_native=True)
+            
+            # if we haven't already overloaded this method
+            if attr not in dir(_LocalTensor) or getattr(_LocalTensor, attr) is None:
+                setattr(_LocalTensor, attr, new_attr)
+    
+    def _hook_SyftTensor(self, tensor_type):
+        
+        self._add_registration_to___init__(_SyftTensor)        
+        
+        for attr in self.to_auto_overload[tensor_type]:
+
+            lit = getattr(tensor_type, attr)
+            passer = utils.pass_method_args(lit)
+            new_attr = self._forward_call_to_children(passer, attr)
+            
+            # if we haven't already overloaded this method
+            if attr not in dir(_SyftTensor) or getattr(_SyftTensor, attr) is None:
+                setattr(_SyftTensor, attr, new_attr)
+    
+    def _forward_call_to_children(hook_self, method, attr, call_native=False):
+        """
+        Wrapper overloading partialmethod objects of Torch object
+        methods.  Compiles command, checks for Tensors and Variables in
+        the args/kwargs, determines locations of all Tensors and
+        Variables involved in computation, and handles the computation
+        accordingly.
+        """
+
+        @functools.wraps(method)
+        def method_router(self, *args, **kwargs):
+            """
+            This is a routing function. If self is a local
+            tensor (data stored locally), then it executes
+            the call locally. If self is a remote tensor, it
+            executes a call to a remote worker.
+            """
+            results = list()
+            if(call_native):
+                for child in self.children:
+                    results.append(getattr(child, "native_"+attr)(*args, **kwargs))
+            else:
+                for child in self.children:
+                    results.append(getattr(child, attr)(*args, **kwargs))
+
+            for result in results:
+                if(type(result) in torch.tensorvar_types and (not hasattr(result, 'owner'))):
```

Comment:
What is the role of `and (not hasattr(result, 'owner'))):` ?
Isn't it `owner` -> `owners` ?
Assumingly, every element of type in `torch.tensor_types` should have an attribute owners thanks to the hook, right?

Reply:
Ah - yes - I'm going to change tensors from having multiple owners to having single owners, where owner is less about who has control of the tensor and more about which Worker object is managing it.